addons/sale_mrp/tests/test_sale_mrp_kit_bom.py

```diff
@@ -210,3 +210,71 @@ def test_qty_delivered_with_bom(self):
 
         # Checks the delivery amount (must be 10).
         self.assertEqual(so.order_line.qty_delivered, 10)
+
+
+    def test_qty_delivered_with_bom_using_kit(self):
+        """Check the quantity delivered, when one product is a kit
+        and his bom uses another product that is also a kit"""
+
+        self.kitA = self._create_product('Kit A', 'product', 0.00)
+        self.kitB = self._create_product('Kit A', 'product', 0.00)
+        self.compA = self._create_product('ComponentA', 'product', 0.00)
+        self.compB = self._create_product('ComponentB', 'product', 0.00)
+
+        # Create BoM for KitB
+        bom_product_formA = Form(self.env['mrp.bom'])
+        bom_product_formA.product_id = self.kitB
+        bom_product_formA.product_tmpl_id = self.kitB.product_tmpl_id
+        bom_product_formA.product_qty = 1.0
+        bom_product_formA.type = 'phantom'
+        with bom_product_formA.bom_line_ids.new() as bom_line:
+            bom_line.product_id = self.compA
+            bom_line.product_qty = 1
+        with bom_product_formA.bom_line_ids.new() as bom_line:
+            bom_line.product_id = self.compB
+            bom_line.product_qty = 1
+        self.bomA = bom_product_formA.save()
+
+        # Create BoM for KitA
+        bom_product_formB = Form(self.env['mrp.bom'])
+        bom_product_formB.product_id = self.kitA
+        bom_product_formB.product_tmpl_id = self.kitA.product_tmpl_id
+        bom_product_formB.product_qty = 1.0
+        bom_product_formB.type = 'phantom'
+        with bom_product_formB.bom_line_ids.new() as bom_line:
+            bom_line.product_id = self.compA
+            bom_line.product_qty = 1
+        with bom_product_formB.bom_line_ids.new() as bom_line:
+            bom_line.product_id = self.kitB
+            bom_line.product_qty = 1
+        self.bomB = bom_product_formB.save()
+
+
+        self.customer = self.env['res.partner'].create({
+            'name': 'customer',
+        })
+
+        so = self.env['sale.order'].create({
+            'partner_id': self.customer.id,
+            'order_line': [
+                (0, 0, {
+                    'name': self.kitA.name,
+                    'product_id': self.kitA.id,
+                    'product_uom_qty': 1.0,
+                    'product_uom': self.kitA.uom_id.id,
+                    'price_unit': 1,
+                    'tax_id': False,
+                })],
+        })
+        so.action_confirm()
+
+        self.assertTrue(so.picking_ids)
+        self.assertEqual(so.order_line.qty_delivered, 0)
+
+        picking = so.picking_ids
+        for move_line in picking.move_lines:
+            move_line.quantity_done = 1
+        picking.button_validate()
```

Comment:
Here it needs some changes
Before the fix, because two SM are merged, `picking.move_lines` contains two SMs: one for 1 x Compo B and one for 2 x Compo A. Therefore, setting the done quantity to 1 does no respect the use case idea. By the way, when calling `button_validate`, it returns an action to know if you want a backorder, it does not call `action_done` (and thus the issue you are fixing is not highlighted)

So, either change the for loop:
```py
    move_line.quantity_done = move_line.product_uom_qty
```
or
```py
picking = so.picking_ids
action = picking.button_validate()
wizard = self.env[action['res_model']].browse(action['res_id'])
wizard.process()
```
The second option needs your components to be consumable (actually, you could set the 4 products as consumable: it makes more sense with kits and here we don't care about the components stocks)